File: src/windows/AgentHaloPaths.cs
using System;
using System.IO;

namespace CodexHalo
{
    /// <summary>
    /// Canonical paths under %USERPROFILE%\.agent-halo for layout version 2.
    /// Production code must resolve runtime agent data through this type.
    /// Legacy helpers exist only for migration / scrub decisions.
    /// </summary>
    internal static class AgentHaloPaths
    {
        public const int LayoutVersion = 2;

        public static string Root(string userProfile = null)
        {
            return Path.Combine(ResolveUserProfile(userProfile), ".agent-halo");
        }

        public static string LayoutVersionFile(string userProfile = null)
        {
            return Path.Combine(Root(userProfile), ".layout-version");
        }

        public static string BinDirectory(string userProfile = null)
        {
            return Path.Combine(Root(userProfile), "bin");
        }

        public static string StateDirectory(string userProfile = null)
        {
            return Path.Combine(Root(userProfile), "state");
        }

        public static string LogsDirectory(string userProfile = null)
        {
            return Path.Combine(Root(userProfile), "logs");
        }

        public static string CacheDirectory(string userProfile = null)
        {
            return Path.Combine(Root(userProfile), "cache");
        }

        public static string ClaudeStatusLog(string userProfile = null)
        {
            return Path.Combine(LogsDirectory(userProfile), "claude-status.jsonl");
        }

        public static string GrokStatusLog(string userProfile = null)
        {
            return Path.Combine(LogsDirectory(userProfile), "grok-status.jsonl");
        }

        public static string ClaudeContextsDirectory(string userProfile = null)
        {
            return Path.Combine(CacheDirectory(userProfile), "claude-contexts");
        }

        public static string UsageSnapshots(string userProfile = null)
        {
            return Path.Combine(CacheDirectory(userProfile), "usage-snapshots-v1.json");
        }

        // MARK: Legacy (migration / scrub only)

        public static string LegacyClaudeStatusLog(string userProfile = null)
        {
            return Path.Combine(Root(userProfile), "claude-code-status.jsonl");
        }

        public static string LegacyGrokStatusLog(string userProfile = null)
        {
            return Path.Combine(Root(userProfile), "grok-build-status.jsonl");
        }

        public static string LegacyClaudeContextsDirectory(string userProfile = null)
        {
            return Path.Combine(Root(userProfile), "claude-code-contexts");
        }

        public static string LegacyClaudeContextFile(string userProfile = null)
        {
            return Path.Combine(Root(userProfile), "claude-code-context.json");
        }

        public static string LegacyUsageSnapshotsInAgentHalo(string userProfile = null)
        {
            return Path.Combine(Root(userProfile), "usage-snapshots-v1.json");
        }

        public static string LegacyUsageSnapshotsInAppData()
        {
            return Path.Combine(SettingsStorage.AppDirectory, "usage-snapshots-v1.json");
        }

        public static string LegacyAgentHaloHookExe(string userProfile = null)
        {
            return Path.Combine(Root(userProfile), "AgentHaloHook.exe");
        }

        private static string ResolveUserProfile(string userProfile)
        {
            if (!String.IsNullOrEmpty(userProfile))
            {
                return userProfile;
            }
            return Environment.GetFolderPath(Environment.SpecialFolder.UserProfile);
        }
    }
}

File: src/macos/Sources/AgentHaloCore/AgentHaloPaths.swift
import Foundation

/// Canonical paths under `~/.agent-halo` for layout version 2.
///
/// Production code must resolve runtime paths through this type (or inject an
/// instance built for a test home). Legacy properties exist only for migration
/// and scrub decisions — business readers/writers must not open them.
public struct AgentHaloPaths: Sendable, Equatable {
    public static let layoutVersion: Int = 2

    public let root: URL
    public let homeDirectory: URL

    public var layoutVersionFile: URL {
        root.appendingPathComponent(".layout-version")
    }

    public var binDirectory: URL {
        root.appendingPathComponent("bin", isDirectory: true)
    }

    public var stateDirectory: URL {
        root.appendingPathComponent("state", isDirectory: true)
    }

    public var logsDirectory: URL {
        root.appendingPathComponent("logs", isDirectory: true)
    }

    public var cacheDirectory: URL {
        root.appendingPathComponent("cache", isDirectory: true)
    }

    public var statusHook: URL {
        binDirectory.appendingPathComponent("status-hook")
    }

    public var statuslineProxy: URL {
        binDirectory.appendingPathComponent("statusline-proxy")
    }

    public var statuslineOriginalCommand: URL {
        stateDirectory.appendingPathComponent("statusline-original-command")
    }

    public var claudeStatusLog: URL {
        logsDirectory.appendingPathComponent("claude-status.jsonl")
    }

    public var grokStatusLog: URL {
        logsDirectory.appendingPathComponent("grok-status.jsonl")
    }

    public var piStatusLog: URL {
        logsDirectory.appendingPathComponent("pi-status.jsonl")
    }

    public var claudeContextsDirectory: URL {
        cacheDirectory.appendingPathComponent("claude-contexts", isDirectory: true)
    }

    public var usageSnapshots: URL {
        cacheDirectory.appendingPathComponent("usage-snapshots-v1.json")
    }

    // MARK: - Legacy (migration / scrub only)

    public var legacyStatusHook: URL {
        root.appendingPathComponent("claude-code-status-hook")
    }

    public var legacyStatuslineProxy: URL {
        root.appendingPathComponent("claude-code-statusline-proxy")
    }

    public var legacyStatuslineOriginalCommand: URL {
        root.appendingPathComponent("claude-code-statusline-original-command")
    }

    public var legacyClaudeStatusLog: URL {
        root.appendingPathComponent("claude-code-status.jsonl")
    }

    public var legacyGrokStatusLog: URL {
        root.appendingPathComponent("grok-build-status.jsonl")
    }

    public var legacyClaudeContextsDirectory: URL {
        root.appendingPathComponent("claude-code-contexts", isDirectory: true)
    }

    public var legacyClaudeContextFile: URL {
        root.appendingPathComponent("claude-code-context.json")
    }

    public var legacyUsageSnapshots: URL {
        root.appendingPathComponent("usage-snapshots-v1.json")
    }

    public init(homeDirectory: URL = FileManager.default.homeDirectoryForCurrentUser) {
        self.homeDirectory = homeDirectory
        self.root = homeDirectory.appendingPathComponent(".agent-halo", isDirectory: true)
    }
}
